{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BioE 231 HW7"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulating the data\n",
    "Generate binary files with different percentages of zeros and ones and protein sequence. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-18-84f3a086bec3>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mpercentages\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;36m0.5\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0.6\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0.7\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0.8\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0.9\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mpercentage\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mpercentages\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0mboolean\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrandom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mchoice\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m8\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;36m1024\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;36m1024\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mpercentage\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mpercentage\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m     \u001b[0mboolean\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpackbits\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mboolean\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'zeros_'\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpercentage\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m'p'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "percentages = [0.5,0.6,0.7,0.8,0.9,1]\n",
    "for percentage in percentages:\n",
    "    boolean = np.random.choice([0, 1], size=8*1024*1024*100, replace=True, p=[percentage, 1-percentage]) \n",
    "    boolean = np.packbits(boolean)\n",
    "    name = 'zeros_'+str(int(percentage*100))+'p'\n",
    "    # write this data to a file in home directory\n",
    "    open(name, 'wb').write(boolean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "nt_sequence = np.random.choice(['A','T','C','G'], size=100000000, replace=True, p=[0.25,0.25,0.25,0.25])\n",
    "open('nt_sequence.fa','w').write(''.join(nt_sequence))\n",
    "\n",
    "pt_codes = ['A','R','N','D','C','E','Q','G','H','I','L','K','M','F','P','S','T','W','Y','V']\n",
    "pt_sequence = np.random.choice(pt_codes, size=100000000, replace=True, p=[0.05]*20)\n",
    "open('pt_sequence.fa','w').write(''.join(pt_sequence))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compressing the data\n",
    "For the files generated above, run gzip, bzip, pbzip2 and ArithmeticCompress. The time it took to run each file using the four methods are recorded below. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 50% zeros: \n",
    "```sh\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time gzip -k zeros_50p\n",
    "\n",
    "real    0m3.502s\n",
    "user    0m3.413s\n",
    "sys     0m0.088s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time bzip2 -k zeros_50p\n",
    "\n",
    "real    0m16.694s\n",
    "user    0m16.581s\n",
    "sys     0m0.113s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time ArithmeticCompress zeros_50p zeros_50p.art\n",
    "\n",
    "real    0m41.227s\n",
    "user    0m40.962s\n",
    "sys     0m0.229s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time pbzip2 -k zeros_50p\n",
    "\n",
    "real    0m1.500s\n",
    "user    0m39.210s\n",
    "sys     0m0.943s\n",
    "```\n",
    "### 60% zeros: \n",
    "```sh\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time gzip -k zeros_60p\n",
    "\n",
    "real    0m4.258s\n",
    "user    0m4.145s\n",
    "sys     0m0.112s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time bzip2 -k zeros_60p\n",
    "\n",
    "real    0m15.766s\n",
    "user    0m15.629s\n",
    "sys     0m0.136s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time ArithmeticCompress zeros_60p zeros_60p.art\n",
    "\n",
    "real    0m41.246s\n",
    "user    0m41.124s\n",
    "sys     0m0.120s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time pbzip2 -k zeros_60p\n",
    "\n",
    "real    0m1.357s\n",
    "user    0m35.473s\n",
    "sys     0m0.825s\n",
    "```\n",
    "### 70% zeros: \n",
    "```sh\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time gzip -k zeros_70p\n",
    "\n",
    "real    0m6.221s\n",
    "user    0m6.048s\n",
    "sys     0m0.168s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time bzip2 -k zeros_70p\n",
    "\n",
    "real    0m13.790s\n",
    "user    0m13.709s\n",
    "sys     0m0.081s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time ArithmeticCompress zeros_60p zeros_70p.art\n",
    "\n",
    "real    0m2.597s\n",
    "user    0m2.579s\n",
    "sys     0m0.017s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time pbzip2 -k zeros_70p\n",
    "\n",
    "real    0m1.179s\n",
    "user    0m30.356s\n",
    "sys     0m0.773s\n",
    "```\n",
    "### 80% zeros: \n",
    "```sh\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time gzip -k zeros_80p\n",
    "\n",
    "real    0m13.435s\n",
    "user    0m13.259s\n",
    "sys     0m0.117s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time bzip2 -k zeros_80p\n",
    "\n",
    "real    0m12.001s\n",
    "user    0m11.881s\n",
    "sys     0m0.121s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time ArithmeticCompress zeros_80p zeros_80p.art\n",
    "\n",
    "real    0m35.714s\n",
    "user    0m35.366s\n",
    "sys     0m0.325s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time pbzip2 -k zeros_80p\n",
    "\n",
    "real    0m0.959s\n",
    "user    0m23.869s\n",
    "sys     0m0.817s\n",
    "```\n",
    "### 90% zeros: \n",
    "```sh\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time gzip -k zeros_90p\n",
    "\n",
    "real    0m19.363s\n",
    "user    0m19.192s\n",
    "sys     0m0.141s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time bzip2 -k zeros_90p\n",
    "\n",
    "real    0m10.649s\n",
    "user    0m10.564s\n",
    "sys     0m0.084s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time ArithmeticCompress zeros_90p zeros_90p.art\n",
    "\n",
    "real    0m29.345s\n",
    "user    0m29.192s\n",
    "sys     0m0.152s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time pbzip2 -k zeros_90p\n",
    "\n",
    "real    0m0.753s\n",
    "user    0m18.753s\n",
    "sys     0m0.607s\n",
    "```\n",
    "### 100% zeros: \n",
    "```sh\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time gzip -k zeros_100p\n",
    "\n",
    "real    0m0.725s\n",
    "user    0m0.657s\n",
    "sys     0m0.064s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time bzip2 -k zeros_100p\n",
    "\n",
    "real    0m1.050s\n",
    "user    0m0.998s\n",
    "sys     0m0.052s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time ArithmeticCompress zeros_100p zeros_100p.art\n",
    "\n",
    "real    0m14.816s\n",
    "user    0m14.751s\n",
    "sys     0m0.065s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time pbzip2 -k zeros_100p\n",
    "\n",
    "real    0m0.110s\n",
    "user    0m1.813s\n",
    "sys     0m0.112s\n",
    "```\n",
    "### nucleotide sequence: \n",
    "```sh\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time gzip -k nt_sequence.fa\n",
    "\n",
    "real    0m12.297s\n",
    "user    0m12.244s\n",
    "sys     0m0.053s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time bzip2 -k nt_sequence.fa\n",
    "\n",
    "real    0m9.521s\n",
    "user    0m9.464s\n",
    "sys     0m0.057s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time ArithmeticCompress nt_sequence.fa nt_sequence.fa.art\n",
    "\n",
    "real    0m21.686s\n",
    "user    0m21.593s\n",
    "sys     0m0.093s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time pbzip2 -k nt_sequence.fa\n",
    "\n",
    "real    0m0.653s\n",
    "user    0m15.823s\n",
    "sys     0m0.530s\n",
    "```\n",
    "### protein sequence: \n",
    "```sh\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time gzip -k pt_sequence.fa\n",
    "\n",
    "real    0m4.246s\n",
    "user    0m4.150s\n",
    "sys     0m0.096s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time bzip2 -k pt_sequence.fa\n",
    "\n",
    "real    0m9.958s\n",
    "user    0m9.893s\n",
    "sys     0m0.065s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time ArithmeticCompress pt_sequence.fa pt_sequence.fa.art\n",
    "\n",
    "real    0m30.016s\n",
    "user    0m29.943s\n",
    "sys     0m0.073s\n",
    "\n",
    "be131-05@meowth:~/LAB7/bio131test/LAB7$ time pbzip2 -k pt_sequence.fa\n",
    "\n",
    "real    0m0.766s\n",
    "user    0m18.678s\n",
    "sys     0m0.705s\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Runtime and input size for each file \n",
    "\n",
    "| File name | gzip(s)| bzip2(s)| pbzip2(s)| Arithmetic compress(s)|Input size| \n",
    "| --------- |:---------:| :---------:|:----------:|:-------:|:--------:|\n",
    "| fifty     | 3.512     | 17.310     | 1.513      | 42.064|105 MB | \n",
    "| sixty     | 4.279     | 16.075     | 1.375      | 41.306|105 MB | \n",
    "| seventy   | 5.971     | 14.276     | 1.140      | 39.535|105 MB | \n",
    "| eighty    | 13.490    | 12.111     | 0.935      | 35.460|105 MB | \n",
    "| ninety    | 18.786    | 10.789     | 0.740      | 29.547|105 MB | \n",
    "| full      | 0.715     | 1.02       | 0.096      | 14.838|105 MB | \n",
    "| nt seq    | 12.156    | 9.581      | 0.664      | 21.903|100 MB | \n",
    "| pt seq    | 4.275     | 10.415     | 0.730      | 28.880|100 MB | \n",
    "\n",
    "### Output file size using different compression algorithms\n",
    "| File name | gzip(.gz) | bzip2(.bz2)| pbzip2(.bz2)| Arithmetic compress(in s)|\n",
    "| --------- |:---------:| :---------:|:----------: |:------------------------:|\n",
    "| fifty     | 105 MB    |  105 MB          | 105 MB      | 105 MB  | \n",
    "| sixty     | 102 MB    | 105MB      | 105 MB      | 102 MB  | \n",
    "| seventy   | 93.6 MB   | 99.8 MB     | 99.8 MB     | 102 MB  | \n",
    "| eighty    | 81.2 MB   | 86.6 MB     | 86.7 MB     | 75.7 MB | \n",
    "| ninety    | 58.7 MB   | 61.2 MB     | 61.2 MB     | 49.2 MB | \n",
    "| full      | 102 KB    | 113 B       | 5.62 KB     | 1.03 KB | \n",
    "| nt seq    | 29.2 MB   | 27.3 MB      | 27.3 MB     | 25 MB   |\n",
    "| pt seq    | 60.6 MB   | 55.3 MB     | 55.3 MB     | 54 MB   |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions\n",
    "- Which algorithm achieves the best level of compression on each file type?\n",
    "    - For nucleotide sequence and protein sequence files, Arithmetic compress works the best. For files containing only zeros and ones, Arithmetic compress also achieves better level of compression compared to other methods. As the percentage of zero increases. At high percentages of zeros (50% & 40%), the advantage of using Arithmetic compression is not as obvious. \n",
    "    \n",
    "    \n",
    "- Which algorithm is the fastest?\n",
    "    - pbzip2 is the fastest algorithms among all.\n",
    "    \n",
    "    \n",
    "- What is the difference between bzip2 and pbzip2? Do you expect one to be faster and why?\n",
    "    - pbzip2 is a parallel implementation of the bzip2 block-sorting file compressor. bzip2 compresses data in blocks and uses the Burrowsâ€“Wheeler transform to convert frequently-recurring character sequences into strings of identical letters. It then applies move-to-front transform and Huffman coding. pbzip2 is a modified version of bzip2 that supports multi-threading, giving almost linear speed improvements. In this case, pbzip2 is expected to work faster, which is confirmed by the results summarized in the first table.\n",
    "    \n",
    "    \n",
    "- How does the level of compression change as the percentage of zeros increases? Why does this happen?\n",
    "    - As the percentage of zeros increases, the level of compression also increases. This happens since fewer bits are required to encode 0.\n",
    "    \n",
    "    \n",
    "- What is the minimum number of bits required to store a single DNA base?\n",
    "    - The minimum number of bits required to store a single DNA base is $log_{2}(4) = 2 bits$, since there are four possible bases.\n",
    "    \n",
    "    \n",
    "- What is the minimum number of bits required to store an amino acid letter?\n",
    "    - The minimum number of bits required to store an amino acid letter is $log_{2}(20) = 4.32bits$, since there are 20 amino acids.\n",
    "    \n",
    "    \n",
    "- In your tests, how many bits did gzip and bzip2 actually require to store your random DNA and protein sequences?\n",
    "    - For a random DNA sequence 100 million long, the bits required should be $2 bits/nt * 100*10^6 nt = 200 * 10^6 bits = 200 MB$ in theory. Similarly, the bits required for encoding a protein sequence with the same length should be $4 bits/AA * 100*10^6 AA = 400 * 10^6 bits = 400 MB$ in theory if we approximate 4.32 bits to 4 bits per amino acid. gzip is able to achieve 29.2 MB and 60.6 MB for DNA and protein sequences respectively, while bzip2 requires 27.3 MB and 55.3 MB in reality.\n",
    "    \n",
    "    \n",
    "- Are gzip and bzip2 performing well on DNA and proteins?\n",
    "    - In turns of level of compression, gzip and bzip2 give similar performances compared to pbzip2 and Arithmetic Compress.  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compressing real data \n",
    "Find  the nucleic acid sequences of gp120 homologs from at least 10 different HIV isolates and concatenate them together into a single multi-FASTA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "Entrez.email = 'vivianfu@berkeley.edu'\n",
    "handle = Entrez.esearch(db='nucleotide',\n",
    "                        term = 'gp120',\n",
    "                        sort = 'relevance',\n",
    "                        idtype = 'acc'\n",
    "                       )                      \n",
    "\n",
    "record = Entrez.read(handle)\n",
    "ID = record['IdList']\n",
    "seqs = ''\n",
    "\n",
    "for i in ID:\n",
    "    handle = Entrez.efetch(db = 'nucleotide', id = i, rettype = 'fasta', retmode = 'text')\n",
    "    record = next(SeqIO.parse(handle, 'fasta'))\n",
    "    seqs += str(record.seq)\n",
    "\n",
    "open('gp120.fa','w').write(seqs)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- A priori, do you expect to achieve better or worse compression here than random data? Why?\n",
    "    - It should achieve better compression than random data. The actual nucleoacid sequence is not random, which should make compression more efficient.\n",
    "    \n",
    "    \n",
    "### Using compress algorithms:\n",
    "\n",
    "    be131-05@meowth:~/LAB7$ time gzip -k gp120.fa\n",
    "    real    0m0.002s\n",
    "    user    0m0.002s\n",
    "    sys     0m0.000s\n",
    "    \n",
    "    bzip2:\n",
    "    be131-05@meowth:~/LAB7$ time bzip2 -k gp120.fa\n",
    "    real    0m0.006s\n",
    "    user    0m0.006s\n",
    "    sys     0m0.000s\n",
    "    \n",
    "    be131-05@meowth:~/LAB7$ time ArithmeticCompress gp120.fa gp120.fa.art\n",
    "    real    0m0.010s\n",
    "    user    0m0.010s\n",
    "    sys     0m0.000s\n",
    "    \n",
    "    be131-05@meowth:~/LAB7/bio131test/LAB7$ rm gp120.bz2\n",
    "    \n",
    "    be131-05@meowth:~/LAB7/bio131test/LAB7$ time pbzip2 -k gp120.fa\n",
    "\n",
    "    real    0m0.003s\n",
    "    user    0m0.003s\n",
    "    sys     0m0.000s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "                          \n",
    "- input and output file sizes for gp120 and random DNA sequence\n",
    "\n",
    "| file name            \t| input size \t| output size  gzip \t| output size  bzip2 \t| output size  pbzip2 \t| output size  Arithmetic Compress \t|\n",
    "|----------------------\t|------------\t|-------------------\t|--------------------\t|---------------------\t|----------------------------------\t|\n",
    "| gp120                \t| 8.44 kB    \t| 1.88 kB           \t| 1.91 kB            \t| 1.91 kB             \t| 3.04 kB                          \t|\n",
    "| random DNA  sequence \t| 100 MB     \t| 29.2 MB           \t| 27.3 MB            \t| 27.3 MB             \t| 25 MB                            \t|\n",
    "\n",
    "\n",
    "\n",
    "                             "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Compression ratio for gp120 and random DNA sequence\n",
    "           \n",
    "           \n",
    "                             \n",
    "|       file name               \t| gzip  \t| bzip2 \t| pbzip2 \t| Arithmetic Compress \t|\n",
    "|----------------------\t|-------\t|-------\t|--------\t|---------------------\t|\n",
    "| gp120                \t| 77.7% \t|77.4% \t| 77.4%  \t| 64%               \t|\n",
    "| random DNA  sequence \t| 71.8% \t| 72.7% \t| 72.7%  \t| 75%                 \t|"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- How does the compression ratio of this file compare to random data?\n",
    "\n",
    "    - For the most part, the compression ratio is higher for real DNA sequence as displayed in the table above, which confirms with the prior assumption that real data should achieve better compression. However, the random sequence achieves better compression than real sequence in the case of Arithmetic Compress. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estimating compression of 1000 terabytes\n",
    "1000 TB of data to begin with. Among which:\n",
    "- 800 TB Genomes and plasmids with high similarity: \n",
    "    - Use gzip. gzip can achieves the highest compression ratio for real DNA data as seen above. Assume that the compression ratio is 77.7% (same as gp120), the reduction in data in this case is $800{TB}* 77.7\\% = 621.6{TB}$.\n",
    "\n",
    "\n",
    "- 100 TB Proteins: \n",
    "    - use pbzip2. pbzip2 and bzip2 work equally well for compressing protein sequences. Although ArithmeticCompress work slightly better based on the output file size results for random protein sequence, the difference is very small. Since pbzip2 works significantly faster than ArithmeticCompress, we choose pbzip2 to compress protein data assuming the compression ratio to be $(100{MB}-55.3{MB})/100{MB} = 44.7\\%$. The reduction in data for protein sequence is $100{TB}*44.7\\%=44.7{TB}$.\n",
    "\n",
    "\n",
    "- 100 TB Random binary data: \n",
    "    - use ArithmeticCompress. Based on the results for files with random percentages of zeros, ArithmeticCompress seems to produce the best level of compression. For a completely random binary file, the compression seems to stay in the range of 0% ~ 10% based on observation from the previous results. We assume the compression to be 10% here. The reduction in data is $100{TB}*10\\%=10{TB}$.\n",
    "    \n",
    "The overall reduction in data is $621.6+44.7+10 = 676.3{TB}$. This translates to 67.63 % reduction in data. If 500 dollars bonus is rewarded for every 1% reduction in data, the total bonus would be $(67.63\\%/1\\%)*{$500} = {$33,815}$."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
