{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LAB8\n",
    "## Running SPAdes\n",
    "- Why do we expect short reads to produce a more fragmented assembly than long reads? \n",
    "    - The assembly produced by using short reads would be more fragmented because more contigs are required. The assembly would be more broken up and fragmented. Longer reads, on the other hand, would be more coherent and are unlikely to result in a broken up assembly.  \n",
    "    \n",
    "- Why does a single-molecule sequencing like PacBio have a higher error rate than Illumina?\n",
    "    - PacBio(single molecule sequencing) tends to have a higher error rate than short reads because the reads are longer, which makes the process of examining the data inefficient. Consequently, Pacbio is expected to be less accurate than illumina. One way to compensate for the high error rate is to get multiple reads from the same template.  \n",
    "\n",
    "The SPAdes command is shown below:\n",
    "\n",
    "    spades.py -1 /data/lab8/illumina_reads_R1.fastq -2 /data/lab8/illumina_reads_R2.fastq --pacbio /data/lab8/pacbio_reads.fastq -t 1 -o /home/be131-05/LAB8/illumina_concatenated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LAB9\n",
    "## Generate assembly statistics\n",
    "Running the assembly-stats program:\n",
    "\n",
    "    be131-05@meowth:~/LAB8/illumina_concatenated$ assembly-stats contigs.fasta scaffolds.fasta\n",
    "    \n",
    "    stats for contigs.fasta\n",
    "    sum = 6678635, n = 176, ave = 37946.79, largest = 414008\n",
    "    N50 = 234690, n = 11\n",
    "    N60 = 202142, n = 14\n",
    "    N70 = 154232, n = 18\n",
    "    N80 = 103367, n = 23\n",
    "    N90 = 81286, n = 30\n",
    "    N100 = 128, n = 176\n",
    "    N_count = 0\n",
    "    Gaps = 0\n",
    "    -------------------------------------------------------------------------------\n",
    "    stats for scaffolds.fasta\n",
    "    sum = 6678655, n = 174, ave = 38383.07, largest = 414008\n",
    "    N50 = 234690, n = 11\n",
    "    N60 = 202142, n = 14\n",
    "    N70 = 154232, n = 18\n",
    "    N80 = 103367, n = 23\n",
    "    N90 = 81821, n = 30\n",
    "    N100 = 128, n = 174\n",
    "    N_count = 20\n",
    "    Gaps = 2\n",
    "\n",
    "The results are summarized in the following table. \n",
    "\n",
    "|  -              | Contigs   | Scaffolds   |\n",
    "|:---------------:|:---------:|:-----------:|\n",
    "| Total Length    |  6678635  |   6678655   |\n",
    "| Number          |  176      |   174       |\n",
    "| N50             |   234690  |   234690    |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Why is N50 is useful statistic to calculate? Why not just list the mean or median contig length? In your assembly, are scaffolds longer than contigs, or are scaffolds approximately equal to contigs?\n",
    "    - N50 describes the completeness of the genome assembly and shows the distribution of the contig lengths. The N50 is defined as the minimum contig length needed to cover 50% of the genome. It means, half of the genome sequence is in contigs larger than or equal the N50 contig size. It is useful in genomic statistic and is more meaningful than mean/median because it has greater weight on longer contigs. In reality, some contigs might be too short and thus are not very useful since it's not desirable to fill up an assembly with many overly short sequences. Researchers sometimes choose to exclude the overly short contigs from an assembly. In this case, using a mean or median to describe an assembly would be inaccurate and misleading.  \n",
    "    - The total length of scaffolds are approximately equal to contigs, which means that the assembler has no additional information that could be used to determine the order and orientation of contigs in a scaffold."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extra credit\n",
    "Extract the coverage from each FASTA header and plot a histogram of coverage for all contigs in your assembly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[7.740928, 10.006665, 7.539874, 9.114046, 9.241629, 8.918606, 8.126053, 8.706677, 7.455302, 7.139043, 8.64251, 7.910328, 6.973549, 8.264545, 7.418715, 8.846256, 8.57711, 9.584186, 8.773953, 7.418265, 7.113839, 8.039997, 10.081286, 7.785886, 7.854012, 8.146401, 7.136158, 8.555536, 7.722379, 9.311795, 8.238385, 9.074883, 7.372789, 7.420957, 6.46957, 8.116505, 7.61281, 7.993083, 7.424264, 7.895898, 7.608171, 7.601905, 7.314205, 8.45544, 14.119172, 10.363252, 5.239566, 7.030905, 5.388827, 0.708005, 9.625, 6.539275, 0.427495, 4.630975, 0.934599, 0.812236, 0.803797, 6.890063, 0.900634, 0.898649, 8.990244, 0.98005, 0.97878, 0.929919, 0.937669, 0.834688, 0.9375, 0.923706, 0.937158, 0.958217, 0.958217, 0.944134, 1.002801, 0.72191, 0.974576, 0.977337, 0.864023, 0.982906, 0.985714, 0.933526, 0.742775, 0.94186, 1.008772, 0.918129, 1.008798, 0.692082, 0.970588, 1.02071, 0.946588, 0.952381, 1.026866, 0.855856, 1.638554, 1.039275, 1.021148, 0.740181, 1.045455, 0.936364, 0.681818, 0.772036, 0.929664, 0.642202, 2.276074, 1.0, 0.92638, 0.769939, 1.0, 0.969231, 0.963077, 0.849231, 1.055556, 1.006173, 0.987654, 0.728395, 1.065015, 0.650155, 0.975155, 0.928571, 1.071651, 1.053125, 0.98125, 0.9279, 0.782334, 0.702532, 1.092063, 0.879365, 0.869427, 0.859425, 0.84345, 1.092949, 0.877814, 0.836013, 1.093548, 0.909677, 0.841935, 1.135922, 1.11039, 0.714286, 0.644951, 0.918301, 0.859477, 0.787582, 0.84918, 0.836066, 1.082237, 1.033003, 0.838284, 0.688742, 0.930233, 6.848, 6.79803, 11.80597, 6.305085, 3.566265, 13.716312, 2.442029, 25.487805, 13.710526, 17.648148, 37.012821, 4.408451, 8.758065, 6.785714, 2.042553, 13.767442, 2.692308, 8.297297, 4.84375, 17.055556, 14.785714, 7.071429, 5.714286, 4.214286, 17.0, 1536.0, 150.0]\n",
      "\n",
      "The mean coverage of contigs is, 14.047804437500002\n"
     ]
    }
   ],
   "source": [
    "# the coverage at the fifth position in the first line of fasta\n",
    "from Bio import SeqIO\n",
    "import matplotlib.pyplot as plt \n",
    "import numpy as np\n",
    "records = SeqIO.parse('contigs.fasta','fasta')\n",
    "coverages = []\n",
    "for record in records:\n",
    "    coverage = float(record.id.split('_')[5])\n",
    "    coverages.append(coverage)\n",
    "print(coverages)\n",
    "print()\n",
    "print('The mean coverage of contigs is,',np.mean(coverages))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can see that the last two coverages are much larger compared to the rest of the covergages, we exlude these two when plotting the histogram."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYIAAAEKCAYAAAAfGVI8AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDMuMC4wLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvqOYd8AAAEuVJREFUeJzt3X20ZXVdx/H3xxmVB1MgJiRAB400FvmAo+lCDcWSpASLUFe6yCi0fCJrBboqqZUtLFPJykKx0ExAQCExFZ/TVcDwYDyMCgoqyMOUImImgt/+OL8bt/HO3HPHu8859/7er7Xuumf/zt7nfGevuedzfr+992+nqpAk9ete0y5AkjRdBoEkdc4gkKTOGQSS1DmDQJI6ZxBIUucMAknqnEEgSZ0zCCSpc2unXcA4dt9991q/fv20y5CkFeWSSy75z6pat9h6KyII1q9fz8aNG6ddhiStKEm+NM56Dg1JUucMAknqnEEgSZ0zCCSpcwaBJHXOIJCkzhkEktQ5g0CSOmcQSFLnVsSVxT+I9Secv2D79ScdNuFKJGk22SOQpM4ZBJLUOYNAkjpnEEhS5wwCSeqcQSBJnTMIJKlzBoEkdc4gkKTOGQSS1DmDQJI6ZxBIUucMAknqnEEgSZ0zCCSpcwaBJHXOIJCkzhkEktQ5g0CSOmcQSFLnDAJJ6pxBIEmdMwgkqXMGgSR1btAgSPLbSa5KcmWSdyXZIcm+SS5Mcm2SM5LcZ8gaJEnbNlgQJNkLeBmwoaoOANYAzwFeC7yhqn4M+DpwzFA1SJIWN/TQ0FpgxyRrgZ2Am4CnAme1508Djhi4BknSNgwWBFV1I/A64MuMAuAbwCXAbVV1V1vtBmCvoWqQJC1uyKGhXYHDgX2BHwV2Bg5dwvbHJtmYZOPmzZsHqlKSNOTQ0NOA66pqc1V9FzgHOAjYpQ0VAewN3LjQxlV1SlVtqKoN69atG7BMSerbkEHwZeDxSXZKEuAQ4GrgY8CRbZ2jgXMHrEGStIghjxFcyOig8KXAFe29TgGOB16R5Frgh4FTh6pBkrS4tYuvsv2q6tXAq7do/iLwuCHfV5I0Pq8slqTOGQSS1DmDQJI6ZxBIUucMAknqnEEgSZ0zCCSpcwaBJHXOIJCkzhkEktQ5g0CSOmcQSFLnDAJJ6pxBIEmdMwgkqXMGgSR1ziCQpM4ZBJLUOYNAkjpnEEhS5wwCSeqcQSBJnTMIJKlzBoEkdc4gkKTOGQSS1DmDQJI6ZxBIUucMAknqnEEgSZ0zCCSpcwaBJHXOIJCkzhkEktQ5g0CSOmcQSFLnBg2CJLskOSvJZ5NsSvKEJLsluSDJNe33rkPWIEnatqF7BCcDH6iqhwOPBDYBJwAfqar9gI+0ZUnSlAwWBEkeADwZOBWgqu6sqtuAw4HT2mqnAUcMVYMkaXFD9gj2BTYDf5/ksiRvTbIzsEdV3dTWuRnYY6GNkxybZGOSjZs3bx6wTEnq21hBkOQnt+O11wIHAm+uqkcD32KLYaCqKqAW2riqTqmqDVW1Yd26ddvx9pKkcYzbI/ibJBcl+a025DOOG4AbqurCtnwWo2C4JcmeAO33rUuqWJK0rMYKgqp6EvArwD7AJUn+KcnPLLLNzcBXkjysNR0CXA2cBxzd2o4Gzt2ewiVJy2PtuCtW1TVJfh/YCPwl8OgkAV5VVedsZbOXAu9Mch/gi8ALGIXPmUmOAb4EHPWD/AMkST+YsYIgySMYfYgfBlwA/EJVXZrkR4F/AxYMgqq6HNiwwFOHbF+5kqTlNm6P4E3AWxl9+//2XGNVfbX1EiRJK9S4QXAY8O2quhsgyb2AHarqv6vqHYNVJ0ka3LhnDX0Y2HHe8k6tTZK0wo0bBDtU1R1zC+3xTsOUJEmapHGD4FtJDpxbSPIY4NvbWF+StEKMe4zgOODdSb4KBHgg8OzBqpIkTcxYQVBVFyd5ODB3cdjnquq7w5UlSZqUsS8oAx4LrG/bHJiEqnr7IFVJkiZm3AvK3gE8FLgcuLs1F2AQSNIKN26PYAOwf5stVJK0iox71tCVjA4QS5JWmXF7BLsDVye5CPjOXGNVPXOQqiRJEzNuEJw4ZBGSpOkZ9/TRTyR5MLBfVX04yU7AmmFLkyRNwri3qvwNRncY+7vWtBfw3qGKkiRNzrgHi18MHATcDqOb1AA/MlRRkqTJGTcIvlNVd84tJFnLVm46L0laWcYNgk8keRWwY7tX8buBfx6uLEnSpIwbBCcAm4ErgBcC7we8M5kkrQLjnjX0PeAt7UeStIqMO9fQdSxwTKCqHrLsFUmSJmopcw3N2QH4ZWC35S9HkjRpYx0jqKr/mvdzY1W9kdEN7SVJK9y4Q0MHzlu8F6MewlLuZSBJmlHjfpj/xbzHdwHXA0ctezWSpIkb96yhpwxdiCRpOsYdGnrFtp6vqtcvTzmSpElbyllDjwXOa8u/AFwEXDNEUZKkyRk3CPYGDqyqbwIkORE4v6qeN1RhkqTJGHeKiT2AO+ct39naJEkr3Lg9grcDFyV5T1s+AjhtmJIkSZM07llDr0nyL8CTWtMLquqy4cqSJE3KuENDADsBt1fVycANSfYdqCZJ0gSNe6vKVwPHA69sTfcG/nGooiRJkzNuj+BZwDOBbwFU1VeBHxqqKEnS5IwbBHdWVdGmok6y87hvkGRNksuSvK8t75vkwiTXJjkjyX2WXrYkabmMGwRnJvk7YJckvwF8mPFvUvNyYNO85dcCb6iqHwO+DhwzbrGSpOU37jTUrwPOAs4GHgb8YVW9abHtkuzNaLrqt7blAE9trwWjU1CPWHrZkqTlsujpo0nWAB9uE89dsMTXfyPwe9xzPOGHgduq6q62fAOw1xJfU5K0jBbtEVTV3cD3kjxgKS+c5OeBW6vqku0pLMmxSTYm2bh58+bteQlJ0hjGvbL4DuCKJBfQzhwCqKqXbWObg4BnJnkGo9tb3h84mdFxhrWtV7A3cONCG1fVKcApABs2bPi++yVLkpbHuEFwTvsZW1W9knbdQZKDgd+tql9J8m7gSOB04Gjg3KW8riRpeW0zCJI8qKq+XFXLOa/Q8cDpSf4EuAw4dRlfW5K0RIv1CN4LHAiQ5Oyq+qXteZOq+jjw8fb4i8Djtud1JEnLb7GDxZn3+CFDFiJJmo7FgqC28liStEosNjT0yCS3M+oZ7Nge05arqu4/aHWSpMFtMwiqas2kCpEkTcdS7kcgSVqFDAJJ6pxBIEmdMwgkqXMGgSR1ziCQpM4ZBJLUOYNAkjpnEEhS5wwCSeqcQSBJnTMIJKlzBoEkdc4gkKTOGQSS1DmDQJI6ZxBIUucMAknqnEEgSZ0zCCSpcwaBJHXOIJCkzhkEktQ5g0CSOmcQSFLnDAJJ6pxBIEmdMwgkqXMGgSR1ziCQpM4ZBJLUOYNAkjo3WBAk2SfJx5JcneSqJC9v7bsluSDJNe33rkPVIEla3JA9gruA36mq/YHHAy9Osj9wAvCRqtoP+EhbliRNyWBBUFU3VdWl7fE3gU3AXsDhwGlttdOAI4aqQZK0uIkcI0iyHng0cCGwR1Xd1J66GdhjK9scm2Rjko2bN2+eRJmS1KXBgyDJ/YCzgeOq6vb5z1VVAbXQdlV1SlVtqKoN69atG7pMSerWoEGQ5N6MQuCdVXVOa74lyZ7t+T2BW4esQZK0bUOeNRTgVGBTVb1+3lPnAUe3x0cD5w5VgyRpcWsHfO2DgOcDVyS5vLW9CjgJODPJMcCXgKMGrEGStIjBgqCqPgVkK08fMtT79mT9Cecv2H79SYdNuBJJK5lXFktS5wwCSeqcQSBJnTMIJKlzBoEkdc4gkKTOGQSS1DmDQJI6ZxBIUucMAknq3JBzDWlKnHpC0lLYI5CkzhkEktQ5g0CSOucxgo547EDSQuwRSFLnDAJJ6pxBIEmdMwgkqXMGgSR1ziCQpM55+ugKsLXTPiVpOdgjkKTOGQSS1DmDQJI6ZxBIUucMAknqnEEgSZ0zCCSpcwaBJHXOC8o0k/cpmMWapNXKHoEkdc4ewRaW65vo9kwLMWvfdv1WLvXBHoEkdc4egbpkb0e6x1R6BEkOTfK5JNcmOWEaNUiSRiYeBEnWAH8N/BywP/DcJPtPug5J0sg0hoYeB1xbVV8ESHI6cDhw9SSLWOrB3EncE2Cl33dgEsMtQ7+HQ0b3cF9MzrT39TSGhvYCvjJv+YbWJkmaglTVZN8wORI4tKp+vS0/H/ipqnrJFusdCxzbFh8GfG7Mt9gd+M9lKndI1rl8VkKNsDLqXAk1gnWO68FVtW6xlaYxNHQjsM+85b1b2/9TVacApyz1xZNsrKoN21/eZFjn8lkJNcLKqHMl1AjWudymMTR0MbBfkn2T3Ad4DnDeFOqQJDGFHkFV3ZXkJcAHgTXA26rqqknXIUkamcoFZVX1fuD9A738koeTpsQ6l89KqBFWRp0roUawzmU18YPFkqTZ4lxDktS5VRUEK2XqiiTXJ7kiyeVJNk67HoAkb0tya5Ir57XtluSCJNe037tOs8ZW00J1npjkxrY/L0/yjCnXuE+SjyW5OslVSV7e2mdqf26jzlnbnzskuSjJZ1qdf9Ta901yYft7P6OdfDJrNf5Dkuvm7ctHTavGbVk1Q0Nt6orPAz/D6CK1i4HnVtVEr1geR5LrgQ1VNTPnQSd5MnAH8PaqOqC1/Rnwtao6qQXrrlV1/AzWeSJwR1W9bpq1zUmyJ7BnVV2a5IeAS4AjgF9lhvbnNuo8itnanwF2rqo7ktwb+BTwcuAVwDlVdXqSvwU+U1VvnrEaXwS8r6rOmkZd41pNPYL/m7qiqu4E5qau0Biq6pPA17ZoPhw4rT0+jdGHxFRtpc6ZUlU3VdWl7fE3gU2Mrp6fqf25jTpnSo3c0Rbv3X4KeCow9wE71f25jRpXhNUUBCtp6ooCPpTkknYF9azao6puao9vBvaYZjGLeEmS/2hDR1MfwpqTZD3waOBCZnh/blEnzNj+TLImyeXArcAFwBeA26rqrrbK1P/et6yxqub25WvavnxDkvtOscStWk1BsJI8saoOZDQD64vbcMdMq9EY4qx+w3kz8FDgUcBNwF9Mt5yRJPcDzgaOq6rb5z83S/tzgTpnbn9W1d1V9ShGMxE8Dnj4lEv6PlvWmOQA4JWMan0ssBsw1aHVrVlNQTDW1BWzoKpubL9vBd7D6D/2LLqljSPPjSffOuV6FlRVt7Q/wu8Bb2EG9mcbJz4beGdVndOaZ25/LlTnLO7POVV1G/Ax4AnALknmroWamb/3eTUe2obfqqq+A/w9M7Qv51tNQbAipq5IsnM7MEeSnYGfBa7c9lZTcx5wdHt8NHDuFGvZqrkP1+ZZTHl/tgOHpwKbqur1856aqf25tTpncH+uS7JLe7wjoxNCNjH6sD2yrTbV/bmVGj87L/jD6BjGTP6tr5qzhgDaaW5v5J6pK14z5ZK+T5KHMOoFwOjK7n+ahTqTvAs4mNFsibcArwbeC5wJPAj4EnBUVU31QO1W6jyY0TBGAdcDL5w3Fj9xSZ4I/CtwBfC91vwqRuPvM7M/t1Hnc5mt/fkIRgeD1zD68npmVf1x+1s6ndGQy2XA89o371mq8aPAOiDA5cCL5h1UnhmrKggkSUu3moaGJEnbwSCQpM4ZBJLUOYNAkjpnEEhS5wwCrVpJHpjk9CRfaNN5vD/Jj0+7LmnWGARaldoFPO8BPl5VD62qxzC63H/Z5/eZd3WrtCIZBFqtngJ8t6r+dq6hqj4DfCrJnye5MqN7QjwboPUcDptbt80jf2SbSOzPk1zcJg57YXv+4CT/muQ84OrW9t7W87hq/mSCSY5J8vk2X/1bkvxVa1+X5Oz22hcnOai1//S8+esvm7sSXRqK32S0Wh3AaH79Lf0io6tmH8no6uSLk3wSOIPRPPzntylKDgF+EzgG+EZVPbbNHPnpJB9qr3UgcEBVXdeWf62qvtamGLg4ydnAfYE/aOt+E/go8Jm2/snAG6rqU0keBHwQ+Angd4EXV9Wn24Rw/7NM+0RakEGg3jwReFdV3c1oErhPMJoZ8l+Ak9uH/aHAJ6vq20l+FnhEkrk5bR4A7AfcCVw0LwQAXpbkWe3xPm29BwKfmJtKIsm7gbnjFE8D9h+NYgFw//bB/2ng9UneyejGKzcs8z6Q/h+DQKvVVdwzIdmiqup/knwceDrwbEZz2MBojpiXVtUH56+f5GDgW1ssPw14QlX9d3utHRZ523sBj6+qLb/xn5TkfOAZjHogT6+qz477b5GWymMEWq0+Ctx3i7H6RwC3Ac9uY//rgCcDF7VVzgBeADwJ+EBr+yDwm226ZpL8eJs1dksPAL7eQuDhwONb+8XATyfZtR1U/qV523wIeOm8+h7Vfj+0qq6oqte27Wdu7n2tLvYItCpVVbVhmjcmOZ7ROPv1wHHA/RiN0xfwe1V1c9vsQ8A7gHPb7U4B3gqsBy5tZyJtZuFbIn4AeFGSTcDngH9vddyY5E8Zhc3XgM8C32jbvAz46yT/wehv8ZOM7nF7XJKnMJoR9CpGw1bSYJx9VBpYkvu1m5qvZXRK69uq6j2LbSdNikND0vBOzOhetlcC1zG6z4M0M+wRSFLn7BFIUucMAknqnEEgSZ0zCCSpcwaBJHXOIJCkzv0v4SSM2zZu/GEAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(coverages[:-2],50)\n",
    "plt.xlabel('Coverages')\n",
    "plt.ylabel('Frequency')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Is coverage uniformly distributed? Does it look Gaussian? Is it bimodal or trimodal?\n",
    "    - The coverage is not a uniform distribution or a Gaussian distribution. It is a bimodal distribution with one peak at 1 and a second peak at ~7. \n",
    "- What explains the presence of contigs with coverage an integer multiple of the mean coverage?\n",
    "    - The mean coverage is determined to be 14. We do not observe coverages that are multiple of the mean coverage here. If we do, then it means the regions in genome are duplicated. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identify the taxon from which your genome originated\n",
    "To identify your genome, you must\n",
    "\n",
    "1. scan over the entire genome to locate copies of the 16S ribosomal RNA gene, \n",
    "2. extract the 16S rRNA genes from your assembly, and \n",
    "3. compare these genes to a database of known 16S rRNA genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- First, run the rna_hmm3.py program on your assembled contigs to locate rRNA genes. \n",
    "     \n",
    "         be131-05@meowth:~/LAB8$ rna_hmm3.py -i contigs.fasta -o contigs.gff\n",
    "    \n",
    "- The output gff content is included here. Delete all lines within other than those containing 16S_rRNA genes.    \n",
    "    \n",
    "        ##seq_name\tmethod\tfeature\tstart\tend\tevalue\tstrand\tframe\tattribute\n",
    "        NODE_23_length_103367_cov_10.081286\trna_hmm3\trRNA\t35531\t37075\t0\t+\tNA\t16S_rRNA\n",
    "        NODE_18_length_154232_cov_9.584186\trna_hmm3\trRNA\t124542\t126086\t0\t+\tNA\t16S_rRNA\n",
    "        NODE_5_length_327223_cov_9.241629\trna_hmm3\trRNA\t46286\t47830\t0\t+\tNA\t16S_rRNA\n",
    "        NODE_6_length_298784_cov_8.918606\trna_hmm3\trRNA\t13926\t15470\t0\t+\tNA\t16S_rRNA\n",
    "        NODE_4_length_332509_cov_9.114046\trna_hmm3\trRNA\t137012\t138556\t0\t-\tNA\t16S_rRNA\n",
    "        NODE_2_length_378948_cov_10.006665\trna_hmm3\trRNA\t348500\t350044\t0\t+\tNA\t16S_rRNA\n",
    "    \n",
    "    \n",
    "- Next, use bedtools getfasta to extract nucleic acid sequences of the 16S rRNA genes from\n",
    "your assembly. You will need to specify the path to contigs.fasta with -fi and the path to the\n",
    "GFF file you obtained above, with -bed. The output will be in FASTA format.\n",
    "\n",
    "        be131-05@meowth:~/LAB8$ bedtools getfasta -fi contigs.fasta -bed contigs.gff > 16S_rRNA.fasta\n",
    "\n",
    "- Using the Ribosomal Database Project’s SeqMatch tool, the sequences in the output gff files are examined. One of the results is included here. \n",
    "\n",
    "        ##seq_name\t\n",
    "        NODE_23_length_103367_cov_10.081286: \n",
    "        \n",
    "        rootrank Root (20) (match sequences) \n",
    "        domain Bacteria (20) \n",
    "          phylum \"Proteobacteria\" (20) \n",
    "             class Gammaproteobacteria (20) \n",
    "                order Pseudomonadales (20) \n",
    "                   family Pseudomonadaceae (20) \n",
    "                      genus Pseudomonas (20) \n",
    "                        S000407513     not_calculated 0.984 1186  Pseudomonas sp. MFY32; AY331343\n",
    "                        S000891983     not_calculated 0.984 1296  Pseudomonas sp. II-39; DQ977710\n",
    "                        S001098708     not_calculated 0.985 1234  Pseudomonas lini; REG21; EU647700\n",
    "                        S001098710     not_calculated 0.986 1243  Pseudomonas mediterranea; REG60; EU647702\n",
    "                        S001795896     not_calculated 0.984 1278  Pseudomonas sp. Lc10-2; GU244354\n",
    "                        S002958599     not_calculated 0.989 1337  Pseudomonas fluorescens; hp13; JN637318\n",
    "                        S003278370     not_calculated 1.000 1270  Pseudomonas frederiksbergensis; rp3; JF906070\n",
    "                        S003278371     not_calculated 1.000 1270  Pseudomonas frederiksbergensis; rp13; JF906071\n",
    "                        S003278373     not_calculated 0.984 1269  Pseudomonas mandelii; rp23; JF906073\n",
    "                        S003290716     not_calculated 0.990 1251  Pseudomonas arsenicoxydans; MaBP1; JQ317810\n",
    "                        S003290717     not_calculated 0.984 1250  Pseudomonas mandelii; McBPA4; JQ317811\n",
    "                        S003290718     not_calculated 0.984 1250  Pseudomonas mandelii; McBRA2; JQ317812\n",
    "                        S003302217     not_calculated 0.989 1303  Pseudomonas lini; 40; JX122159\n",
    "                        S003616324     not_calculated 0.989 1236  Pseudomonas mandelii; BJC15-C35; JX469391\n",
    "                        S003616553     not_calculated 0.989 1228  Pseudomonas mandelii; BJC16-C1; JX483782\n",
    "                        S003745408     not_calculated 0.986 1258  Pseudomonas mandelii; Tibetlhz-57; JX827234\n",
    "                        S003789558     not_calculated 0.985 1298  Pseudomonas sp. Aza24; JQ977627\n",
    "                        S004484092     not_calculated 0.984 1241  Pseudomonas mandelii; Ge03; KR088370\n",
    "                        S004493454     not_calculated 0.987 1306  Pseudomonas fluorescens; SE02(2); KT239467\n",
    "                        S004534171     not_calculated 0.984 1246  bacterium C0074; KP177382\n",
    "\n",
    "This tool is used for all 16S rRNA sequences that we obtained previously. Although they map to different species, they all match to a common genus: Pseudomonas. This conclusion is reached since the top 20 matches of each input match to an identical genus. \n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genome annotation\n",
    "\n",
    "Upload the genome to two remote annotation services that will perform\n",
    "the annotation automatically: RAST and BASys.\n",
    "\n",
    "- Annotation from BASYs: https://www.basys.ca/server3/basys/cache/3f767cbac6f990dc1d56ab58e2f7686c/maps/index.html\n",
    "- Annotation from RAST: http://rast.nmpdr.org/seedviewer.cgi?page=Organism&organism=6666666.379375"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Research and write-up\n",
    "\n",
    "- CRISPR-Cas9 is so hot right now. Does your genome have a CRISPR system? Can you determine where the spacer sequences originated from?\n",
    "\n",
    "\n",
    "    - CRISPR-Cas9: Clustered regularly interspaced short palindromic repeats (CRISPR) and their associated genes (cas) are essential components of an adaptive immune system that protects bacteria and archaea from viral infection. CRISPR systems contain two components: a guide RNA (gRNA or sgRNA) and a CRISPR-associated endonuclease (Cas protein). The gRNA is a short synthetic RNA composed of a scaffold sequence necessary for Cas-binding and a user-defined ∼20 nucleotide spacer that defines the genomic target to be modified. Thus, one can change the genomic target of the Cas protein by simply changing the target sequence present in the gRNA. Reference: *https://www.addgene.org/crispr/guide/*\n",
    "    \n",
    "    \n",
    "    - From this research paper (*https://www.nature.com/articles/cr20176*), we can see that Pseudomonas aeruginosa(PA) has a CRISPR system. The spacer sequences in the CRISPR system are mostly complementary to sequences located on its bacteriophage genome (the virus that  parasitizes PA). Some spacer sequences do contain mismatches. Mismatches between PA spacer sequence and corresponding sequence on viral genome was found to prevent the formation of biofilm by causing DNA damage and cell death. Spacer sequences are genetics remnants of previous invaders, which allow PA to detect and destroy bacteriophage if they return.  \n",
    "    \n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
