{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BioE 231 HW4 \n",
    "Four enzymes were selected in each of the three pathways using the KEGG pathway map. They are store in the corresponding pathway in a dictionary. The three species are also stored inside a list called species. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lists of organisms, pathways and enzymes established...\n",
      "gene id, enzyme encoded sequence found..\n",
      "entries for gene table constructed...\n",
      "entries for enzyme table constructed...\n",
      "Gene table is displayed here: \n",
      "    Gene_ID     Name                                        Description  \\\n",
      "0         1     DLAT               dihydrolipoamide S-acetyltransferase   \n",
      "1         2     IDH1    isocitrate dehydrogenase (NADP(+)) 1, cytosolic   \n",
      "2         3     COMT                       catechol-O-methyltransferase   \n",
      "3         4    PHGDH                     phosphoglycerate dehydrogenase   \n",
      "4         5     PGM1                               phosphoglucomutase 1   \n",
      "5         6     GALM                               galactose mutarotase   \n",
      "6         7     PGK1                          phosphoglycerate kinase 1   \n",
      "7         8     ENO1                                          enolase 1   \n",
      "8         9      GPI                      glucose-6-phosphate isomerase   \n",
      "9        10     G6PD                  glucose-6-phosphate dehydrogenase   \n",
      "10       11     FBP1                          fructose-bisphosphatase 1   \n",
      "11       12     DERA                     deoxyribose-phosphate aldolase   \n",
      "12       13      muc                              midline uncoordinated   \n",
      "13       14      Idh                           Isocitrate dehydrogenase   \n",
      "14       15   Nc73EF                           Neural conserved at 73EF   \n",
      "15       16     Mdh1                             Malate dehydrogenase 1   \n",
      "16       17      Pgm                              phosphoglucose mutase   \n",
      "17       18  CG32444                            uncharacterized protein   \n",
      "18       19      Pgk                            Phosphoglycerate kinase   \n",
      "19       20      Eno                                            Enolase   \n",
      "20       21      Pgi                           Phosphoglucose isomerase   \n",
      "21       22       Zw                                    Zwischenferment   \n",
      "22       23      fbp                        fructose-1,6-bisphosphatase   \n",
      "23       24   CG8525                            uncharacterized protein   \n",
      "24       25     aceF  pyruvate dehydrogenase, dihydrolipoyltransacet...   \n",
      "25       26      icd  isocitrate dehydrogenase; e14 prophage attachm...   \n",
      "26       27     sucA  2-oxoglutarate decarboxylase, thiamine triphos...   \n",
      "27       28      mdh               malate dehydrogenase, NAD(P)-binding   \n",
      "28       29      pgm                                 phosphoglucomutase   \n",
      "29       30     galM  aldose 1-epimerase; type-1 mutarotase; galacto...   \n",
      "30       31      pgk                            phosphoglycerate kinase   \n",
      "31       32      eno                                            enolase   \n",
      "32       33      pgi                         glucosephosphate isomerase   \n",
      "33       34      zwf                glucose-6-phosphate 1-dehydrogenase   \n",
      "34       35      fbp                      fructose-1,6-bisphosphatase I   \n",
      "35       36     deoC  2-deoxyribose-5-phosphate aldolase, NAD(P)-linked   \n",
      "\n",
      "                                     Organism Chromosome      Start  \\\n",
      "0                                Homo sapiens         11  112024813   \n",
      "1                                Homo sapiens          2  208255142   \n",
      "2                                Homo sapiens         22   19941739   \n",
      "3                                Homo sapiens          1  119711795   \n",
      "4                                Homo sapiens          1   63593275   \n",
      "5                                Homo sapiens          2   38665909   \n",
      "6                                Homo sapiens          X   78104168   \n",
      "7                                Homo sapiens          1    8878685   \n",
      "8                                Homo sapiens         19   34353329   \n",
      "9                                Homo sapiens          X  154547585   \n",
      "10                               Homo sapiens          9   94640257   \n",
      "11                               Homo sapiens         12   15911171   \n",
      "12                    Drosophila melanogaster         2L    7426865   \n",
      "13                    Drosophila melanogaster         3L    8360905   \n",
      "14                    Drosophila melanogaster         3L   16954821   \n",
      "15                    Drosophila melanogaster         2L   10396331   \n",
      "16                    Drosophila melanogaster         3L   16128530   \n",
      "17                    Drosophila melanogaster         3L   21636946   \n",
      "18                    Drosophila melanogaster         2L    2753067   \n",
      "19                    Drosophila melanogaster         2L    1729635   \n",
      "20                    Drosophila melanogaster         2R    8914068   \n",
      "21                    Drosophila melanogaster          X   19667251   \n",
      "22                    Drosophila melanogaster         2L   19764135   \n",
      "23                    Drosophila melanogaster         2R   12425482   \n",
      "24  Escherichia coli str. K-12 substr. MG1655       None     125694   \n",
      "25  Escherichia coli str. K-12 substr. MG1655       None    1195122   \n",
      "26  Escherichia coli str. K-12 substr. MG1655       None     758705   \n",
      "27  Escherichia coli str. K-12 substr. MG1655       None    3384267   \n",
      "28  Escherichia coli str. K-12 substr. MG1655       None     713557   \n",
      "29  Escherichia coli str. K-12 substr. MG1655       None     788836   \n",
      "30  Escherichia coli str. K-12 substr. MG1655       None    3072621   \n",
      "31  Escherichia coli str. K-12 substr. MG1655       None    2907940   \n",
      "32  Escherichia coli str. K-12 substr. MG1655       None    4233757   \n",
      "33  Escherichia coli str. K-12 substr. MG1655       None    1936313   \n",
      "34  Escherichia coli str. K-12 substr. MG1655       None    4455608   \n",
      "35  Escherichia coli str. K-12 substr. MG1655       None    4617322   \n",
      "\n",
      "          End                                           Sequence  \\\n",
      "0   112064277                           GDEEKRNRAITARRQHLKSVMLQI   \n",
      "1   208236226  MLKNFKLSKRDSNGSKGRITSADISTPSHDNGSVIKHIKTVPVRYL...   \n",
      "2    19969974  EQLTKCEAFQKLKDLKDYGGVSLPEWVCTAFHTSGYDTQAIVQNND...   \n",
      "3   119744225  GAGATTCGGTGACATAGGTTAGTGAGTGGTGTGCTTTCAGACGTGT...   \n",
      "4    63660244  TTCCCTGCTCGTATGTCTTGAGATTGGGCCTTAAATAATATGCAGC...   \n",
      "5    38734766  GASIPFKLERLQALQDLVHRSLEAGYISPWDGPGNNPVFPVRKPNG...   \n",
      "6    78126826  ATCGTGCTCTACGACTTTGCACACCACAAAGTGTGCGGGGAGCAAA...   \n",
      "7     8861001  DEFHPFIEALLPHVRAFAYTWFNLQARKRKYFKKHEKRMSKDEERA...   \n",
      "8    34402412  ATCGATTTGAGAGGTGATAGGTGATTAGCGGGGGGAGATGAAAAGT...   \n",
      "9   154531389  MKLSLLSTFAAVIIGALALPQGPGGGGSVTCPGGQSTSNSQCCVWF...   \n",
      "10   94603132  GGTGGAGGAACAAAAGTTTTCGTCCAAGATAAAAACCTTCCTACAG...   \n",
      "11   16037380  GATCCACCTGGCTCTGGCTTCAGTTCCTGGTACTGAAGATTTGCAC...   \n",
      "12    7430328  PSSLSAAMIARQQCVRGGPRGFSCGSAIVGGGKRGAFSSVSMSGGA...   \n",
      "13    8356403  TTGTTTGGAGAGTTTGATTCTGGCTCAGGACGAACGCTGGCGGCGT...   \n",
      "14   16968723  GTTAACAGGATATCCGAGCTTATCGGCCCACTCGTTCCCAAACACA...   \n",
      "15   10393773  GRRRRSVQWCAVSQPEATKCFQWQRNMRKVRGPPVSCIKRDSPIQC...   \n",
      "16   16126017  CGTTATTTAAGGTGTTACATAGTTCTATGGAAATAGGGTCTATACC...   \n",
      "17   21639059  MAFELPDLPYKLNALEPHISQETLEYHHGKHHRAYVNKLNKLIEGT...   \n",
      "18    2746879  ATGGAGTTTGGGCTGAGCTGGGTTTTCCTTGTTGCTATTTTAAAAG...   \n",
      "19    1724767  GGTCAGCCCAAGGCCACCCCCTCGGTCACTCTGTTCCTGCCGTCCT...   \n",
      "20    8910899  GGATCCCAGCCTTTCCCCAGCCCGTAGCCCCGGGACCTCCGCGGTG...   \n",
      "21   19672352  EVQLVESGGGLVKPGGSLRLSCAASGFTFSSYSMNWVRQAPGKGLE...   \n",
      "22   19761586                              MQMSPALTCLVLGLTLVFGEG   \n",
      "23   12426974  TCTAGACAGAAGCATTCTCAGATGCTTCATTGGGATGTTTCAATTG...   \n",
      "24     127586  CACCAACAGAAGCTTTTATCATAACTTACGAATAACAGAAATAAAG...   \n",
      "25    1196372  TAAGTACAAATGCTTTTATATTCAGCCCCTGTAAAGCCATCAGATG...   \n",
      "26     761506  GCCTGTGACACCGTGGAACTGGGGGAGGAAAAACTGTACACCGATA...   \n",
      "27    3383329  AGACCTGGTGGTTATAAGCCTATTGTGTTTGTCAATGGACCTGTTA...   \n",
      "28     715197  ACTTACTACTGTGCCCTCTCTCTTTACTGGGGGATACAGGCACCGA...   \n",
      "29     787796  MPAYKYNRVFLIVMDSVGIGEAPDAADFNDEGAHTLGHIAEHMNGL...   \n",
      "30    3071458  GCAAAAGCAATTCCNGTAGTAGCGGCGAGCGAAATGGGAGCAGCCT...   \n",
      "31    2906642  CTCGAGGTCGCGGAGACCCCAAGGCCACGGAAGGACCCGCGTCTCC...   \n",
      "32    4235406  GAAAAACAGACAGCGGAGGACTAAAACAACTAGCCACAACACAACG...   \n",
      "33    1934838  CCATNCTGGTGTTCCGCGGCCCGGCCTGNANGTTCTTCCCAGGGTC...   \n",
      "34    4454610  CAAGTAACATTCTCAGCTTCTAGAAGGTTTTCTCACCAACCCCCAA...   \n",
      "35    4618101  GCGCAAAAGCGTTTGGNGGCAGCACGATGCAGCGTCTTCGGTAGCA...   \n",
      "\n",
      "   Encoded_Enzyme  \n",
      "0     ec:2.3.1.12  \n",
      "1     ec:1.1.1.42  \n",
      "2      ec:1.2.4.2  \n",
      "3     ec:1.1.1.37  \n",
      "4      ec:5.4.2.2  \n",
      "5      ec:5.1.3.3  \n",
      "6      ec:2.7.2.3  \n",
      "7     ec:4.2.1.11  \n",
      "8      ec:5.3.1.9  \n",
      "9     ec:1.1.1.49  \n",
      "10    ec:3.1.3.11  \n",
      "11     ec:4.1.2.4  \n",
      "12    ec:2.3.1.12  \n",
      "13    ec:1.1.1.42  \n",
      "14     ec:1.2.4.2  \n",
      "15    ec:1.1.1.37  \n",
      "16     ec:5.4.2.2  \n",
      "17     ec:5.1.3.3  \n",
      "18     ec:2.7.2.3  \n",
      "19    ec:4.2.1.11  \n",
      "20     ec:5.3.1.9  \n",
      "21    ec:1.1.1.49  \n",
      "22    ec:3.1.3.11  \n",
      "23     ec:4.1.2.4  \n",
      "24    ec:2.3.1.12  \n",
      "25    ec:1.1.1.42  \n",
      "26     ec:1.2.4.2  \n",
      "27    ec:1.1.1.37  \n",
      "28     ec:5.4.2.2  \n",
      "29     ec:5.1.3.3  \n",
      "30     ec:2.7.2.3  \n",
      "31    ec:4.2.1.11  \n",
      "32     ec:5.3.1.9  \n",
      "33    ec:1.1.1.49  \n",
      "34    ec:3.1.3.11  \n",
      "35     ec:4.1.2.4  \n",
      "\n",
      "Enzyme table is displayed here: \n",
      "    enzyme_id                                               name  \\\n",
      "0           1      dihydrolipoyllysine-residue acetyltransferase   \n",
      "1           1      dihydrolipoyllysine-residue acetyltransferase   \n",
      "2           2                   isocitrate dehydrogenase (NADP+)   \n",
      "3           3  oxoglutarate dehydrogenase (succinyl-transferr...   \n",
      "4           4                               malate dehydrogenase   \n",
      "5           5  phosphoglucomutase (alpha-D-glucose-1,6-bispho...   \n",
      "6           5  phosphoglucomutase (alpha-D-glucose-1,6-bispho...   \n",
      "7           6                                 aldose 1-epimerase   \n",
      "8           7                            phosphoglycerate kinase   \n",
      "9           8                          phosphopyruvate hydratase   \n",
      "10          9                      glucose-6-phosphate isomerase   \n",
      "11          9                      glucose-6-phosphate isomerase   \n",
      "12         10          glucose-6-phosphate dehydrogenase (NADP+)   \n",
      "13         11                            fructose-bisphosphatase   \n",
      "14         11                            fructose-bisphosphatase   \n",
      "15         12                     deoxyribose-phosphate aldolase   \n",
      "\n",
      "                                             reaction  \\\n",
      "0   acetyl-CoA + enzyme N6-(dihydrolipoyl)lysine =...   \n",
      "1   acetyl-CoA + enzyme N6-(dihydrolipoyl)lysine =...   \n",
      "2   isocitrate + NADP+ = 2-oxoglutarate + CO2 + NA...   \n",
      "3   2-oxoglutarate + [dihydrolipoyllysine-residue ...   \n",
      "4   (S)-malate + NAD+ = oxaloacetate + NADH + H+ [...   \n",
      "5   alpha-D-glucose 1-phosphate = D-glucose 6-phos...   \n",
      "6   alpha-D-glucose 1-phosphate = D-glucose 6-phos...   \n",
      "7        alpha-D-glucose = beta-D-glucose [RN:R01602]   \n",
      "8   ATP + 3-phospho-D-glycerate = ADP + 3-phospho-...   \n",
      "9   2-phospho-D-glycerate = phosphoenolpyruvate + ...   \n",
      "10  D-glucose 6-phosphate = D-fructose 6-phosphate...   \n",
      "11  D-glucose 6-phosphate = D-fructose 6-phosphate...   \n",
      "12  D-glucose 6-phosphate + NADP+ = 6-phospho-D-gl...   \n",
      "13  D-fructose 1,6-bisphosphate + H2O = D-fructose...   \n",
      "14  D-fructose 1,6-bisphosphate + H2O = D-fructose...   \n",
      "15  2-deoxy-D-ribose 5-phosphate = D-glyceraldehyd...   \n",
      "\n",
      "                                              comment enzyme_commission  \\\n",
      "0   A multimer (24-mer or 60-mer, depending on the...       ec:2.3.1.12   \n",
      "1   A multimer (24-mer or 60-mer, depending on the...       ec:2.3.1.12   \n",
      "2   Requires Mn2+ or Mg2+ for activity. Unlike EC ...       ec:1.1.1.42   \n",
      "3   Contains thiamine diphosphate. It is a compone...        ec:1.2.4.2   \n",
      "4   Also oxidizes some other 2-hydroxydicarboxylic...       ec:1.1.1.37   \n",
      "5   Maximum activity is only obtained in the prese...        ec:5.4.2.2   \n",
      "6   Maximum activity is only obtained in the prese...        ec:5.4.2.2   \n",
      "7   Also acts on L-arabinose, D-xylose, D-galactos...        ec:5.1.3.3   \n",
      "8                                                 N/A        ec:2.7.2.3   \n",
      "9               Also acts on 3-phospho-D-erythronate.       ec:4.2.1.11   \n",
      "10  Also catalyses the anomerization of D-glucose ...        ec:5.3.1.9   \n",
      "11  Also catalyses the anomerization of D-glucose ...        ec:5.3.1.9   \n",
      "12  The enzyme catalyses a step of the pentose pho...       ec:1.1.1.49   \n",
      "13  The animal enzyme also acts on sedoheptulose 1...       ec:3.1.3.11   \n",
      "14  The animal enzyme also acts on sedoheptulose 1...       ec:3.1.3.11   \n",
      "15                                                N/A        ec:4.1.2.4   \n",
      "\n",
      "   pathway_id  \n",
      "0           2  \n",
      "1           1  \n",
      "2           1  \n",
      "3           1  \n",
      "4           1  \n",
      "5           2  \n",
      "6           3  \n",
      "7           2  \n",
      "8           2  \n",
      "9           2  \n",
      "10          2  \n",
      "11          3  \n",
      "12          3  \n",
      "13          2  \n",
      "14          3  \n",
      "15          3  \n",
      "\n",
      "Pathway table is displayed here: \n",
      "   pathway_id               name  \\\n",
      "0           1          TCA cycle   \n",
      "1           2         glycolysis   \n",
      "2           3  pentose phosphate   \n",
      "\n",
      "                                         description  \n",
      "0  The tricarboxylic acid cycle (TCA cycle) is a ...  \n",
      "1  Glycolysis breaks down glucose and forms pyruv...  \n",
      "2  The pentose phosphate pathway is a metabolic p...  \n",
      "\n",
      "Enzyme to gene associative table:\n",
      "    enzyme_id  gene_id\n",
      "0           1        1\n",
      "1           2        2\n",
      "2           3        3\n",
      "3           4        4\n",
      "4           5        5\n",
      "5           6        6\n",
      "6           7        7\n",
      "7           8        8\n",
      "8           9        9\n",
      "9          10       10\n",
      "10         11       11\n",
      "11         12       12\n",
      "12          1       13\n",
      "13          2       14\n",
      "14          3       15\n",
      "15          4       16\n",
      "16          5       17\n",
      "17          6       18\n",
      "18          7       19\n",
      "19          8       20\n",
      "20          9       21\n",
      "21         10       22\n",
      "22         11       23\n",
      "23         12       24\n",
      "24          1       25\n",
      "25          2       26\n",
      "26          3       27\n",
      "27          4       28\n",
      "28          5       29\n",
      "29          6       30\n",
      "30          7       31\n",
      "31          8       32\n",
      "32          9       33\n",
      "33         10       34\n",
      "34         11       35\n",
      "35         12       36\n",
      "\n",
      "Enzyme to pathway associative table:\n",
      "    enzyme_id  pathway_id\n",
      "0           1           2\n",
      "1           1           1\n",
      "2           2           1\n",
      "3           3           1\n",
      "4           4           1\n",
      "5           5           2\n",
      "6           5           3\n",
      "7           6           2\n",
      "8           7           2\n",
      "9           8           2\n",
      "10          9           2\n",
      "11          9           3\n",
      "12         10           3\n",
      "13         11           2\n",
      "14         11           3\n",
      "15         12           3\n"
     ]
    }
   ],
   "source": [
    "import xmltodict\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import sys\n",
    "\n",
    "Entrez.email = 'vivianfu@berkeley.edu'\n",
    "pathways_enzymes = {'TCA cycle': ['ec:2.3.1.12','ec:1.1.1.42', 'ec:1.2.4.2', 'ec:1.1.1.37'],\n",
    "                  'glycolysis': ['ec:5.4.2.2', 'ec:5.1.3.3', 'ec:2.7.2.3', 'ec:4.2.1.11'],\n",
    "                  'pentose phosphate': ['ec:5.3.1.9', 'ec:1.1.1.49', 'ec:3.1.3.11', 'ec:4.1.2.4']}\n",
    "species = ['Homo sapiens','fruit fly','Escherichia coli']\n",
    "\n",
    "print('lists of organisms, pathways and enzymes established...')\n",
    "#######################################################################################\n",
    "# Use entrez database to obtain information of genes for enzymes\n",
    "# The obtained gene IDs, nucleotides are stored in a list of tuples\n",
    "\n",
    "EC_Id_Seq = []\n",
    "for organism in species:\n",
    "    for (pathway,enzymes) in pathways_enzymes.items():\n",
    "        for enzyme in enzymes:            \n",
    "            handle = Entrez.esearch(db='gene',\n",
    "                                    term = organism + ' [ORGN] ' + 'AND ' + enzyme,\n",
    "                                    sort = 'relevance',\n",
    "                                    idtype = 'acc',\n",
    "                                    retmax = 1\n",
    "                                   )                      \n",
    "            \n",
    "            record = Entrez.read(handle)\n",
    "            ID = record['IdList']\n",
    "\n",
    "            for i in ID:\n",
    "                handle = Entrez.efetch(db = 'nucleotide', id = i, rettype = 'gb', retmode = 'text')\n",
    "                record = next(SeqIO.parse(handle, \"genbank\"))\n",
    "                EC_Id_Seq.append((enzyme, ID, str(record.seq)))\n",
    "\n",
    "print('gene id, enzyme encoded sequence found..')\n",
    "#######################################################################################\n",
    "# The inputs of gene table are first organized as dictionaries stored in gene_dicts\n",
    "# Each gene_dict stores the following information:\n",
    "# name, description, organism, nucleotide sequence, chromosome, start, end, strand, translated sequence\n",
    "\n",
    "gene_dicts = []\n",
    "gene_id = 0\n",
    "for (ec,i, seq) in EC_Id_Seq:\n",
    "    gene_dict = {}\n",
    "    handle = Entrez.esummary(db=\"gene\", id=i, retmode=\"xml\") \n",
    "    # convert xml to dictionary\n",
    "    mydict = xmltodict.parse(handle.read()) \n",
    "    # the useful information is in the document summary\n",
    "    DocumentSummary = mydict['eSummaryResult']['DocumentSummarySet']['DocumentSummary'] \n",
    "    \n",
    "    name = DocumentSummary['Name']\n",
    "    description = DocumentSummary['Description']\n",
    "    organism = DocumentSummary['Organism']['ScientificName']\n",
    "    chromosome = DocumentSummary['Chromosome']\n",
    "    start = DocumentSummary['GenomicInfo']['GenomicInfoType']['ChrStart']\n",
    "    end = DocumentSummary['GenomicInfo']['GenomicInfoType']['ChrStop']\n",
    "    sequence = seq\n",
    "    \n",
    "    gene_id += 1\n",
    "    gene_dict['Gene_ID'] = gene_id\n",
    "    gene_dict['Name'] = name\n",
    "    gene_dict['Description'] = description\n",
    "    gene_dict['Organism'] = organism\n",
    "    gene_dict['Chromosome'] = chromosome\n",
    "    gene_dict['Start'] = start\n",
    "    gene_dict['End'] = end\n",
    "    gene_dict['Sequence'] = sequence\n",
    "    gene_dict['Encoded_Enzyme'] = ec     \n",
    "    gene_dicts.append(gene_dict)    \n",
    "\n",
    "print('entries for gene table constructed...')\n",
    "#######################################################################################\n",
    "# The inputs into the enzyme table are obtained from KEGG database\n",
    "# Use KEGG to extract information regarding enzyme and store them in a list of tuples called enzyme_rows\n",
    " \n",
    "from Bio.KEGG import REST\n",
    "from Bio.KEGG import Enzyme\n",
    "\n",
    "enzyme_rows = []\n",
    "enzyme_id = 0\n",
    "\n",
    "# id for these pathways are 1, 2, and 3 respectively\n",
    "pathways = ['Citrate cycle (TCA cycle)','Glycolysis / Gluconeogenesis','Pentose phosphate pathway']\n",
    "\n",
    "for enzymes in pathways_enzymes.values():\n",
    "    for enzyme in enzymes:\n",
    "        enzyme_txt = enzyme + '.txt'\n",
    "        request = REST.kegg_get(enzyme)\n",
    "        open(enzyme_txt, \"w\").write(request.read())\n",
    "        records = Enzyme.parse(open(enzyme_txt))\n",
    "        record = list(records)[0]\n",
    "        name = record.name[0]\n",
    "        reaction = record.reaction[0]\n",
    "        \n",
    "        # some enzymes doesn't have comment in their records\n",
    "        try:\n",
    "            comment = record.comment[0]\n",
    "        except:\n",
    "            comment = 'N/A'\n",
    "            pass\n",
    "        \n",
    "        enzyme_id +=1\n",
    "        \n",
    "        # find the pathways that the enzyme is involved in \n",
    "        for pathway_tuple in record.pathway:\n",
    "            pathway = pathway_tuple[2]\n",
    "            if pathway in pathways:\n",
    "                pathway_id = pathways.index(pathway)+1\n",
    "                enzyme_row = (enzyme_id, name, reaction, comment, enzyme, pathway_id)\n",
    "                enzyme_rows.append(enzyme_row)\n",
    "\n",
    " \n",
    "print('entries for enzyme table constructed...')\n",
    "#######################################################################################\n",
    "# Construct SQL tables\n",
    "\n",
    "import sqlite3\n",
    "conn = sqlite3.connect('my.db')\n",
    "c = conn.cursor()\n",
    "\n",
    "# Allow tables to be regenerated each time the program is run to avoid the 'table already existed' error\n",
    "try:\n",
    "    c.execute(\"DROP TABLE gene_table\")\n",
    "    c.execute(\"DROP TABLE pathway_table\")\n",
    "    c.execute(\"DROP TABLE enzyme_table\")\n",
    "    c.execute(\"DROP TABLE enzyme_pathway_relation_table\")\n",
    "    c.execute(\"DROP TABLE enzyme_gene_relation_table\")\n",
    "    \n",
    "except:\n",
    "    pass \n",
    "\n",
    "#######################################################################################\n",
    "# Construct gene table using the gene entries compiled previously\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE gene_table (Gene_ID, Name TEXT, Description TEXT, Organism TEXT, Chromosome TEXT, Start INT, End INT, Sequence TEXT, Encoded_Enzyme TEXT);\"\"\")\n",
    "\n",
    "gene_rows = []\n",
    "for gene_dict in gene_dicts:\n",
    "    row = tuple(gene_dict.values())\n",
    "    gene_rows.append(row)\n",
    "\n",
    "c.executemany('INSERT INTO gene_table VALUES (?,?,?,?,?,?,?,?,?)', gene_rows)\n",
    "\n",
    "print(\"Gene table is displayed here: \")\n",
    "print (pd.read_sql_query(\"SELECT * FROM gene_table\", conn))\n",
    "print()\n",
    "       \n",
    "#######################################################################################\n",
    "# Construct enzyme table using the enzyme entries compiled previously\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE enzyme_table (enzyme_id INT, name TEXT, reaction TEXT, comment TEXT, enzyme_commission TEXT, pathway_id TEXT);\"\"\")\n",
    "c.executemany('INSERT INTO enzyme_table VALUES (?,?,?,?,?,?)', enzyme_rows)\n",
    "\n",
    "print(\"Enzyme table is displayed here: \")\n",
    "print (pd.read_sql_query(\"SELECT * FROM enzyme_table\", conn))\n",
    "print()\n",
    "\n",
    "#######################################################################################\n",
    "# Construct pathway table: \n",
    "# fields: pathway id, name, description\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE pathway_table (pathway_id, name TEXT, description TEXT);\"\"\")\n",
    "\n",
    "pathway_rows = [(1, 'TCA cycle','The tricarboxylic acid cycle (TCA cycle) is a series of enzyme-catalyzed chemical reactions that form a key part of aerobic respiration in cells. '),\n",
    "                (2, 'glycolysis','Glycolysis breaks down glucose and forms pyruvate with the production of two molecules of ATP. '),\n",
    "                (3, 'pentose phosphate','The pentose phosphate pathway is a metabolic pathway parallel to glycolysis. It generates NADPH and pentoses as well as ribose 5-phosphate.')]\n",
    " \n",
    "c.executemany('INSERT INTO pathway_table VALUES (?,?,?)', pathway_rows)\n",
    "\n",
    "print(\"Pathway table is displayed here: \")\n",
    "print (pd.read_sql_query(\"SELECT * FROM pathway_table\", conn))\n",
    "print()\n",
    "\n",
    "#######################################################################################\n",
    "# Associative tables:\n",
    "\n",
    "# associate table 1: enzyme to gene\n",
    "# enzyme to gene is one to many relationship, one enzyme corresponds to 3 genes across 3 species\n",
    "# therefore, the table has 3 species*12 enzymes = 36 rows in total\n",
    "# fields: enzyme id, gene id\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE enzyme_gene_relation_table (enzyme_id INT, gene_id INT);\"\"\")\n",
    "                                                    \n",
    "c.execute(\"\"\"\n",
    "            SELECT DISTINCT enzyme_table.enzyme_id, gene_table.Gene_Id \n",
    "            from gene_table \n",
    "            left join enzyme_table on enzyme_table.enzyme_commission = gene_table.Encoded_Enzyme\"\"\")\n",
    "\n",
    "enzyme_gene_rows = c.fetchall()\n",
    "c.executemany('INSERT INTO enzyme_gene_relation_table VALUES (?,?)', enzyme_gene_rows)\n",
    "\n",
    "print(\"Enzyme to gene associative table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM enzyme_gene_relation_table\", conn))\n",
    "print()\n",
    "\n",
    "#######################################################################################\n",
    "# associate table 2: enzyme to pathway \n",
    "# enzyme to pathway is many to many relationship, one enzyme can perform in multiple pathways while one pathway \n",
    "# contains many enzymes \n",
    "# fields: enzyme id, pathway id\n",
    "\n",
    "c.execute(\"\"\"CREATE TABLE enzyme_pathway_relation_table (enzyme_id INT, pathway_id INT);\"\"\")\n",
    "c.execute(\"\"\"\n",
    "             SELECT enzyme_table.enzyme_id, enzyme_table.pathway_id\n",
    "             from enzyme_table \n",
    "             \"\"\")\n",
    "\n",
    "enzyme_pathway_rows = c.fetchall()\n",
    "c.executemany('INSERT INTO enzyme_pathway_relation_table VALUES (?,?)', enzyme_pathway_rows)\n",
    "c.execute(\"\"\"select * from enzyme_pathway_relation_table\"\"\")\n",
    "\n",
    "print(\"Enzyme to pathway associative table:\")\n",
    "print (pd.read_sql_query(\"SELECT * FROM enzyme_pathway_relation_table\", conn))\n",
    "\n",
    "\n",
    "conn.commit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
